motoko/icrc2-swap/test/swap/swap.test.ts

```diff
@@ -0,0 +1,261 @@
+
+      // Check deposited balances
+      var balances = await swap(minter).balances();
+      var tokenABalances: Record<string, bigint> = Object.fromEntries(balances[0]);
+      var tokenBBalances: Record<string, bigint> = Object.fromEntries(balances[1]);
+
+      // Alice should have 1 A deposited
+      const alicePrincipal = alice.getPrincipal().toString()
+      expect(tokenABalances[alicePrincipal]).toBe(100_000_000n);
+
+      // Bob should have 1 B deposited
+      const bobPrincipal = bob.getPrincipal().toString()
+      expect(tokenBBalances[bobPrincipal]).toBe(100_000_000n);
+
+      // Do the swap
+      const swapResult = await swap(minter).swap({
+        user_a: alice.getPrincipal(),
+        user_b: bob.getPrincipal(),
+      });
+      expect(swapResult).toHaveProperty("ok");
+
+      // Check deposited balances
+      balances = await swap(minter).balances();
+      tokenABalances = Object.fromEntries(balances[0]);
+      tokenBBalances = Object.fromEntries(balances[1]);
+
+      // Alice should have 1 B deposited
+      expect(tokenBBalances[alicePrincipal]).toBe(100_000_000n);
+
+      // Bob should have 1 A deposited
+      expect(tokenABalances[bobPrincipal]).toBe(100_000_000n);
+
+      // Alice withdraws TokenB
+      const withdrawalA = await swap(alice).withdraw({
+        amount: 100_000_000n - 10_000n,
+        created_at_time : [],
+        fee: [],
+        to : { owner: alice.getPrincipal(), subaccount: [] },
+        memo: [],
+        token : tokenBCanisterId,
+      });
+      expect(withdrawalA).toHaveProperty("ok");
+
+      // Bob withdraws TokenA
+      const withdrawalB = await swap(bob).withdraw({
+        amount: 100_000_000n - 10_000n,
+        created_at_time : [],
+        fee: [],
+        to : { owner: bob.getPrincipal(), subaccount: [] },
+        memo: [],
+        token : tokenACanisterId,
+      });
+      expect(withdrawalB).toHaveProperty("ok");
+
+      // Check the wallet token balances have changed as expected
+      expect(await Promise.all([
+        tokenA(alice).icrc1_balance_of({ owner: alice.getPrincipal(), subaccount: [] }),
+        tokenB(alice).icrc1_balance_of({ owner: alice.getPrincipal(), subaccount: [] }),
+        tokenA(bob).icrc1_balance_of({ owner: bob.getPrincipal(), subaccount: [] }),
+        tokenB(bob).icrc1_balance_of({ owner: bob.getPrincipal(), subaccount: [] }),
+      ])).toEqual([
+        0n,          // Alice should have 0 A
+        99_990_000n, // Alice should have 0.9999 B
+        99_990_000n, // Bob should have 0.9999 A
+        0n,          // Bob should have A B
+      ]);
+    },
+    60_000 // 60 second timeout
+  );
+
+  describe('error handling', () => {
+    test(
+      'deposit with invalid token argument',
+      async () => {
+        // Give alice just enough A
+        const alice = newIdentity();
+
+        // Alice tries to deposit 1 A. This will fail because only 0.5A has
+        // been approved.
```

Comment:
```suggestion
        // Alice tries to deposit a token that does not exist
```

Reply:
Hmm.. I disagree on this one. The token canister exists, but in this scenario Alice hasn't approved any transfers.

Reply:
Oh, right, no I see. It got copied to this test.